### PR TITLE
Feature/ucas api calls

### PR DESCRIPTION
**Pull request by southeo**
### General
This branch introduces new functionality for the Annotation and Digital Media Object endpoints. 

**Important:** New endpoints end in /json. Any existing, modified endpoints have the same behaviour, with optional pagination functionality added. 

Some response objects have been added for selected GET requests:
- JsonApiMeta: includes meta information about the request. Currently has total page count for a given query (possibility to add more meta information as needed here)
- JsonApiLinksFull: a full links object, which includes links to self, prev, next, last, and first pages of a given request 
- JsonApiMetaWrapper: Wrapper object that wraps JsonMeta, JsonApiLinksFull, and a list of JsonApiData 

Where pagination is added to an endpoint, now _optional_ request parameters can be set
- pageNumber: defines the offset, default = 1
- pageSize: defines how many items are returned, default = 10

### Annotation Endpoints

- GET /latest -> added pagination
- GET /latest/json -> pagination, returns JsonApiMetaWrapper object, calls elasticsearch to obtain latest annotations
- GET /all -> pagination introduced, unsorted (calls postgres db, not elasticsearch)
- GET /all/json -> returns JsonApiMetaWrapper object, paginated, calls postgres db not elasticsearch
- GET /creator -> pagination introduced
- GET /creator/json -> paginated call for all annotations by the logged-in user, returns JsonApiMetaWrapper

### Digital Media Object Endpoints
- GET /json -> paginated call for all objects, returns JsonApiMetaWrapper for digital media objects. Under the "attributes" node there is a "specimen" node which contains species information (species name, specimen version number). Only the latest version of the specimen information is returned. 
- GET /json/prefix/postfix -> returns JsonApiWrapper representation of a specified digital media object. The additional information in the JsonApiMetaWrapper (page count, next/previous page, and a list of JsonApiData nodes) doesn't make sense for a single object, so only a JsonApiWrapper is returned (which includes a "self" link and a single JsonApiData node). Returns latest version of the object
- GET /json/prefix/postfix/version -> same as above, but returns a specified version
- GET /prefix/postfix/annotation -> retrieves annotations on a given media object. paginated. Returns JsonApiMetaWrapper object.

### CODE REVIEW

File: src/main/java/eu/dissco/backend/repository/AnnotationRepository.java [hunk 1]
```diff
@@ -21,80 +23,133 @@ public class AnnotationRepository {
...
-        .orderBy(NEW_ANNOTATION.ID, NEW_ANNOTATION.VERSION.desc())
-        .fetchOne(this::mapToAnnotation);
+        .orderBy(NEW_ANNOTATION.ID, NEW_ANNOTATION.VERSION.desc(), NEW_ANNOTATION.CREATED)
+        .limit(pageSize).offset(offset).fetch(this::mapToJsonApiData);
+  }
+
+  public Integer getAnnotationsCountForUser(String userId, int pageSize) {
+    int totalRecords = context.selectCount().from(NEW_ANNOTATION)
+        .where(NEW_ANNOTATION.CREATOR.eq(userId)).fetchOne(0, int.class);
+    return totalRecords / pageSize + ((totalRecords % pageSize == 0) ? 0 : 1);
```
Comment (samleeflang) [inline src/main/java/eu/dissco/backend/repository/AnnotationRepository.java:82]: Would make a separate function for this to remove duplication

File: src/main/java/eu/dissco/backend/repository/AnnotationRepository.java [hunk 2]
```diff
@@ -21,80 +23,133 @@ public class AnnotationRepository {
...
+        .where(NEW_ANNOTATION.ID.eq(id)).fetch(Record1::value1).stream().toList();
   }
 
   public int getAnnotationForUser(String id, String userId) {
-    return context.select(NEW_ANNOTATION.ID)
-        .distinctOn(NEW_ANNOTATION.ID)
-        .from(NEW_ANNOTATION)
-        .where(NEW_ANNOTATION.ID.eq(id))
-        .and(NEW_ANNOTATION.CREATOR.eq(userId))
+    return context.select(NEW_ANNOTATION.ID).distinctOn(NEW_ANNOTATION.ID).from(NEW_ANNOTATION)
```
Comment (samleeflang) [inline src/main/java/eu/dissco/backend/repository/AnnotationRepository.java:150]: I would rather have a bit more overview here and have a newline after each statement

File: src/main/java/eu/dissco/backend/repository/DigitalMediaObjectRepository.java [hunk 3]
```diff
@@ -1,103 +1,225 @@
...
+        .on(NEW_DIGITAL_SPECIMEN.ID.eq(specimenRecentVersions.field(NEW_DIGITAL_SPECIMEN.ID)))
+        .and(NEW_DIGITAL_SPECIMEN.VERSION.eq(specimenRecentVersions.field(maxSpeciesVersion)))
+        .join(NEW_DIGITAL_MEDIA_OBJECT)
+        .on(NEW_DIGITAL_SPECIMEN.ID.eq(NEW_DIGITAL_MEDIA_OBJECT.DIGITAL_SPECIMEN_ID))
+        .where(NEW_DIGITAL_MEDIA_OBJECT.ID.eq(id))
+        .orderBy(NEW_DIGITAL_MEDIA_OBJECT.ID, NEW_DIGITAL_MEDIA_OBJECT.VERSION.desc()).limit(1)
+        .fetchOne(this::mapToJsonApiData);
+  }
+
+  public List<JsonApiData> getAnnotationsOnDigitalMediaObject(String mediaId){
```
Comment (samleeflang) [inline src/main/java/eu/dissco/backend/repository/DigitalMediaObjectRepository.java:58]: I would like to get all calls to the Annotations table in the AnnotationRepository. I would say that the DigitalMediaService -> AnnotationService -> AnnotationRepository

File: src/main/java/eu/dissco/backend/repository/DigitalMediaObjectRepository.java [hunk 4]
```diff
@@ -1,103 +1,225 @@
...
         .orderBy(NEW_DIGITAL_MEDIA_OBJECT.ID, NEW_DIGITAL_MEDIA_OBJECT.VERSION.desc())
         .fetch(this::mapToMultiMediaObject);
   }
 
   public List<Integer> getDigitalMediaVersions(String id) {
-    return context.select(NEW_DIGITAL_MEDIA_OBJECT.VERSION)
-        .from(NEW_DIGITAL_MEDIA_OBJECT)
-        .where(NEW_DIGITAL_MEDIA_OBJECT.ID.eq(id))
-        .fetch(Record1::value1).stream().toList();
+    return context.select(NEW_DIGITAL_MEDIA_OBJECT.VERSION).from(NEW_DIGITAL_MEDIA_OBJECT)
```
Comment (samleeflang) [inline src/main/java/eu/dissco/backend/repository/DigitalMediaObjectRepository.java:83]: Like the earlier version better, couple newlines don't do anyone harm

File: src/main/java/eu/dissco/backend/repository/DigitalMediaObjectRepository.java [hunk 5]
```diff
@@ -1,103 +1,225 @@
...
-      );
-    } catch (JsonProcessingException e) {
-      throw new RuntimeException(e);
-    }
+  public JsonApiData getLatestDigitalMediaObjectByIdJsonResponse(String id) {
+
+    Field<Integer> maxSpeciesVersion = DSL.max(NEW_DIGITAL_SPECIMEN.VERSION)
+        .as("maxSpeciesVersion");
+
+    var specimenRecentVersions = context.select(NEW_DIGITAL_SPECIMEN.ID, maxSpeciesVersion)
```
Comment (samleeflang) [inline src/main/java/eu/dissco/backend/repository/DigitalMediaObjectRepository.java:43]: Not sure what is happening here. It feels like you first make a new table with all latest version of the digital Specimen. This is a very heavy query, and I am not sure why but I might misread it. I would say if you need the digitalMedia of the object you could just query that table. If need specimen info on it you can use the physicalSpecimenId in the digitalMedia to look up more info?

File: src/main/java/eu/dissco/backend/repository/DigitalMediaObjectRepository.java [hunk 6]
```diff
@@ -1,103 +1,225 @@
...
+          dbRecord.get(NEW_DIGITAL_MEDIA_OBJECT.FORMAT),
+          dbRecord.get(NEW_DIGITAL_MEDIA_OBJECT.SOURCE_SYSTEM_ID),
+          mapper.readTree(dbRecord.get(NEW_DIGITAL_MEDIA_OBJECT.DATA).data()),
+          mapper.readTree(dbRecord.get(NEW_DIGITAL_MEDIA_OBJECT.ORIGINAL_DATA).data()));
+    } catch (JsonProcessingException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private JsonApiData mapAnnotationToJsonApiData(Record dbRecord) {
```
Comment (samleeflang) [inline src/main/java/eu/dissco/backend/repository/DigitalMediaObjectRepository.java:170]: If you let annotations handle annotation than you would not have to use this method but can just use the functionality that is already there. Now it is duplicated. If this needs to stay here I would say we move this method to a shared method in a parent class

File: src/main/java/eu/dissco/backend/repository/DigitalMediaObjectRepository.java [hunk 7]
```diff
@@ -1,103 +1,225 @@
...
+      attributeNode.put("mediaUrl",
+          String.valueOf(dbRecord.get(NEW_DIGITAL_MEDIA_OBJECT.MEDIA_URL)));
+      attributeNode.put("format", String.valueOf(dbRecord.get(NEW_DIGITAL_MEDIA_OBJECT.FORMAT)));
+      attributeNode.put("sourceSystemId",
+          String.valueOf(dbRecord.get(NEW_DIGITAL_MEDIA_OBJECT.SOURCE_SYSTEM_ID)));
+      attributeNode.set("data",
+          mapper.readTree(dbRecord.get(NEW_DIGITAL_MEDIA_OBJECT.DATA).data()));
+      attributeNode.set("originalData",
+          mapper.readTree(dbRecord.get(NEW_DIGITAL_MEDIA_OBJECT.ORIGINAL_DATA).data()));
+      if (dbRecord.field(NEW_DIGITAL_SPECIMEN.SPECIMEN_NAME) != null) {
```
Comment (samleeflang) [inline src/main/java/eu/dissco/backend/repository/DigitalMediaObjectRepository.java:213]: When is this null? And why only check on this and not the VERSION?

File: src/main/java/eu/dissco/backend/repository/ElasticSearchRepository.java [hunk 8]
```diff
@@ -8,98 +8,119 @@
...
-        .map(Hit::source)
+        .sort(s -> s.field(f -> f.field(FIELD_CREATED).order(SortOrder.Desc))).from(offset)
+        .size(pageSize).build();
+    return client.search(searchRequest, ObjectNode.class).hits().hits().stream().map(Hit::source)
         .map(this::mapToDigitalSpecimen).toList();
   }
 
+  public List<AnnotationResponse> getLatestAnnotations(int pageNumber, int pageSize)
+      throws IOException {
+    var offset = 0;
```
Comment (samleeflang) [inline src/main/java/eu/dissco/backend/repository/ElasticSearchRepository.java:58]: Some duplication, would move this into a function

File: src/main/java/eu/dissco/backend/repository/ElasticSearchRepository.java [hunk 9]
```diff
@@ -8,98 +8,119 @@
...
+        .sort(s -> s.field(f -> f.field(FIELD_CREATED).order(SortOrder.Desc))).size(10).build();
+    return client.search(searchRequest, ObjectNode.class).hits().hits().stream().map(Hit::source)
         .map(this::mapToAnnotationResponse).toList();
   }
 
+
   private DigitalSpecimen mapToDigitalSpecimen(ObjectNode json) {
+
     var digitalSpecimen = json.get("digitalSpecimen");
-    return new DigitalSpecimen(
```
Comment (samleeflang) [inline src/main/java/eu/dissco/backend/repository/ElasticSearchRepository.java:68]: Can we keep the newlines?

File: src/test/java/eu/dissco/backend/controller/AnnotationControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+    String requestUri = "api/v1/annotations/latest/json";
+    String path = SANDBOX_URI + requestUri;
+    MockHttpServletRequest request = new MockHttpServletRequest();
+    request.setRequestURI(requestUri);
+
+    int pageNumber = 1;
+    int pageSize = 11;
+    int totalPageCount = 100;
+    MockHttpServletRequest r = new MockHttpServletRequest();
+    r.setRequestURI("");
```
Comment (samleeflang) [inline src/test/java/eu/dissco/backend/controller/AnnotationControllerTest.java:139]: How does this work, it looks like it is set two times?

File: src/test/java/eu/dissco/backend/controller/DigitalMediaObjectControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    // Then
+    assertThat(responseReceived.getStatusCode()).isEqualTo(HttpStatus.OK);
+  }
+
+  @Test
+  void testGetAnnotationsById(){
+    // Given
+    int pageNum = 1;
+    int pageSize = 10;
+    int totalPageCount =1;
```
Comment (samleeflang) [inline src/test/java/eu/dissco/backend/controller/DigitalMediaObjectControllerTest.java:155]: Small formatting thing

File: src/test/java/eu/dissco/backend/repository/AnnotationRepositoryIT.java [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+    postAnnotations(annotationsNonTarget);
+
+    List<String> annotationIds = IntStream.rangeClosed(0, pageSize - 1).boxed()
+        .map(Object::toString).toList();
+    List<AnnotationResponse> expectedResponse = new ArrayList<>();
+
+    for (String id : annotationIds) {
+      expectedResponse.add(givenAnnotationResponse(userId, id));
+    }
+    postAnnotations(expectedResponse);
```
Comment (samleeflang) [inline src/test/java/eu/dissco/backend/repository/AnnotationRepositoryIT.java:56]: Small formatting thing, newline

File: src/test/java/eu/dissco/backend/repository/AnnotationRepositoryIT.java [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class AnnotationRepositoryIT extends BaseRepositoryIT {
+
+  private AnnotationRepository repository;
+
+  @BeforeEach
+  void setup() {
+    ObjectMapper mapper = new ObjectMapper().findAndRegisterModules();
```
Comment (samleeflang) [inline src/test/java/eu/dissco/backend/repository/AnnotationRepositoryIT.java:29]: Could use the TestUtils mapper, and add the SerializationInclusion there I think

File: src/test/java/eu/dissco/backend/repository/ElasticSearchRepositoryIT.java [hunk 14]
```diff
@@ -0,0 +1,354 @@
...
+    final CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+    credentialsProvider.setCredentials(AuthScope.ANY,
+        new UsernamePasswordCredentials(ELASTICSEARCH_USERNAME, ELASTICSEARCH_PASSWORD));
+
+    restClient = RestClient.builder(HttpHost.create(container.getHttpHostAddress()))
+        .setHttpClientConfigCallback(httpClientBuilder -> {
+          return httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider);
+        }).build();
+
+    final ObjectMapper mapper = new ObjectMapper();
```
Comment (samleeflang) [inline src/test/java/eu/dissco/backend/repository/ElasticSearchRepositoryIT.java:77]: Could use the TestUtiles mapper?

File: src/test/java/eu/dissco/backend/repository/ElasticSearchRepositoryIT.java [hunk 15]
```diff
@@ -0,0 +1,354 @@
...
+    mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+    mapper.registerModule(new JavaTimeModule());
```
Comment (samleeflang) [inline src/test/java/eu/dissco/backend/repository/ElasticSearchRepositoryIT.java:79]: This is taken care of with the `findAndRegisterModules()`

**Review by samleeflang — COMMENTED** [review]
Looking good, this ups the test coverage massively. Get a couple of comments. Feel free to plan a moment if any are unclear

File: src/main/java/eu/dissco/backend/controller/AnnotationController.java [hunk 16]
```diff
@@ -48,33 +53,69 @@ public ResponseEntity<AnnotationResponse> getAnnotation(
...
+  public ResponseEntity<List<AnnotationResponse>> getAnnotations(
+      @RequestParam(defaultValue = DEFAULT_PAGE_NUM) int pageNumber,
+      @RequestParam(defaultValue = DEFAULT_PAGE_SIZE) int pageSize) {
+    log.info("Received get request for paginated annotations. Page number: {}, page size {}",
+        pageNumber, pageSize);
+    var annotations = service.getAnnotations(pageNumber, pageSize);
+    return ResponseEntity.ok(annotations);
+  }
+
+  @GetMapping(value = "/all/json", produces = MediaType.APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/backend/controller/AnnotationController.java:99]: Good point!

File: src/main/java/eu/dissco/backend/service/AnnotationService.java [hunk 17]
```diff
@@ -59,39 +66,55 @@ public List<AnnotationResponse> getAnnotations(int pageNumber, int pageSize) {
...
       String path) throws IOException {
-    var annotations = elasticRepository.getLatestAnnotationsJsonResponse(pageNumber, pageSize);
-    int totalPageCount = repository.getAnnotationsCountGlobal(pageSize);
-    JsonApiLinksFull linksNode = buildLinksNode(path, pageNumber, pageSize, totalPageCount);
-    JsonApiMeta metaNode = new JsonApiMeta(totalPageCount);
-    return new JsonApiMetaWrapper(annotations, linksNode, metaNode);
+    var annotationsPlusOne = elasticRepository.getLatestAnnotationsJsonResponse(pageNumber, pageSize+1);
+    boolean hasNextPage;
+    List<JsonApiData> annotations;
+    if (annotationsPlusOne.size() > pageSize ){
```
Comment (samleeflang) [inline src/main/java/eu/dissco/backend/service/AnnotationService.java:95]: Could this be a separate method, it seems like it is duplicated multiple times here

File: src/main/java/eu/dissco/backend/service/DigitalMediaObjectService.java [hunk 18]
```diff
@@ -30,14 +30,23 @@ public JsonApiWrapper getDigitalMediaByIdJsonResponse(String id, String path) {
...
       int pageNumber, int pageSize) {
     String mediaId = "https://hdl.handle.net/" + id;
-    var annotations = repository.getAnnotationsOnDigitalMediaObject(mediaId);
-    int totalPageCount = repository.getAnnotationPageCountOnMediaObject(mediaId, pageSize);
-    var metaNode = new JsonApiMeta(totalPageCount);
-    var linksNode = buildLinksNode(path, pageNumber, pageSize, totalPageCount);
-    return new JsonApiMetaWrapper(annotations, linksNode, metaNode);
+    var annotationsPlusOne = repository.getAnnotationsOnDigitalMediaObject(mediaId, pageNumber, pageSize+1);
+
+    boolean hasNextPage;
```
Comment (samleeflang) [inline src/main/java/eu/dissco/backend/service/DigitalMediaObjectService.java:38]: I would probably put this either in a parent or somewhere else as the same lines seem to be used in multiple classes

File: src/test/java/eu/dissco/backend/repository/AnnotationRepositoryIT.java [hunk 19]
```diff
@@ -133,6 +134,7 @@ void testGetAnnotationsForUserJsonResponseSecondPage() {
     assertThat(receivedResponse).hasSize(pageSize);
   }
 
+  /*
```
Comment (samleeflang) [inline src/test/java/eu/dissco/backend/repository/AnnotationRepositoryIT.java:137]: If it is no longer necessary please remove commented out code

File: src/test/java/eu/dissco/backend/repository/DigitalMediaObjectRepositoryIT.java [hunk 20]
```diff
@@ -227,36 +227,20 @@ void testGetAnnotationsOnDigitalMediaObject(){
...
     postAnnotation(annotation, mediaId);
     postMediaObjects(List.of(mediaObject));
+    int pageNumber = 1;
+    int pageSize = 1;
 
     var responseExpected = List.of(givenAnnotationJsonApiData(USER_ID_TOKEN, annotationId));
 
-    var annotations = context.select(NEW_ANNOTATION.asterisk()).from(NEW_ANNOTATION).fetch().stream().toList();
-    var medias = context.select(NEW_DIGITAL_MEDIA_OBJECT.asterisk()).from(NEW_DIGITAL_MEDIA_OBJECT).fetch().stream().toList();
+//    var annotations = context.select(NEW_ANNOTATION.asterisk()).from(NEW_ANNOTATION).fetch().stream().toList();
```
Comment (samleeflang) [inline src/test/java/eu/dissco/backend/repository/DigitalMediaObjectRepositoryIT.java:235]: Can this be removed?

**Review by samleeflang — COMMENTED** [review]
Couple of last comments.

**Review by samleeflang — APPROVED** [review]
Well done!

**Pull request merged**